plugins/volume/org.xen.xapi.storage.zfs-ng/sr.py

```diff
@@ -80,18 +80,22 @@ def destroy(self, dbg, sr):
     def attach(self, dbg, configuration):
         log.debug('{}: SR.attach: config={}'.format(dbg, configuration))
 
-        # ZFS automagically attaches a pool to a mountpoint on
-        # create/boot/etc, so we basically do nothing here but find
-        # this mountpoint
-
-        # FIXME: study how pools are mounted on boot and see if we
-        # could do real a/detach
-
+        pool_name = configuration["zpool"]
+        try:
+            mountpoint = zfsutils.pool_mountpoint(dbg, pool_name)
+            if pool_name in mountpoint:
+                # SR.attach is idempotent. So if the mountpoint is already
+                # mounted just return it
+                return mountpoint
+        except:
+            log.debug('{}: SR.attach: mountpoint for {} not found, import it'.format(dbg, pool_name))
+
+        zfsutils.pool_import(dbg, pool_name)
         return zfsutils.pool_mountpoint(dbg, configuration["zpool"])
 
```

Comment:
This will assume zpool is not imported on any exception from `pool_mountpoint` - is that really correct?  Maybe we can find a more direct test to know if a zpool was already imported?  If not, maybe it would be more correct to rework `pool_mountpoint` to parse the error message itself, returning `None` on `dataset does not exist`, and otherwise reraising?

It may not be obvious what situations we're dealing with the zpool is mounted but `if pool_name in mountpoint` fails: a quick check shows the subsequent `pool_import` will fail with `a pool with that name already exists`

Reply:
"zpool is imported"  you mean. It was my idea that any exception when executing the `pool_mountpoint` will trigger the import + mount of the zpool. I think that to know if it is already mounted the most direct way is to use `zfs get mountpoint`. But yes maybe we need to be more accurate with possible errors returned by the `zpool_mountpoint`.

For me if zpool is mounted then I don't see how `if pool_name in mountpoint` could failed. And if it fails we are not in a good shape so subsequent `pool_import` can fail it is ok I think.

But I think that as you said the `pool_mountpoint` can maybe be improved... I will check what errors are reported by the `zfs get mountpoint`...

Reply:
Also the point if we split the `import` and the `mount` without using the legacy mode I think that you will need to:
1. `zfs import` and check if the SR is listed (if not maybe it is because it is already mounted
    1. If it is not we need to check if it is already mounted
    1. if yes then we need to run `zfs import -R ...` to really mount it (or use legacy but it will need more modifications).
    
So I think that it easier to just try catch to get the mountpoint and if an exception occurs just try to import+mount.

Reply:
> "zpool is imported" you mean

No - sorry I did not explain things well enough.  If `pool_mounpoint` throws an exception (and also if it does not but `pool_name in mountpoint` is false), we go on and call `zfsutils.pool_import`.  If zpool was imported at that time, `pool_import` will fail - and an exception from `pool_mounpoint` is not enough to tell us if the pool is imported.

> For me if zpool is mounted then I don't see how if pool_name in mountpoint could failed.

E.g. users bringing their own zpool could provide it with an arbitrary mountpoint not containing the pool name.  But we should not fail here, just correct the situation (i.e. set zpool config so it does not automount, to be similar to what `zfs create -R ...` produces)

> So I think that it easier to just try catch to get the mountpoint and if an exception occurs just try to import+mount.

"just try to import+mount" won´t work in all cases as noted above - I'm starting to think again that the legacy mode would indeed be much more clear: by forcing us to be explicit, it removes so many edge cases

Reply:
For me users don't mount the zpool. The just create it and we will import it and mount it where we want. It shouldn't be already mounted.

Reply:
So if users provide and zpool and don't mount it we will be able to mount it where we want (`zfs import -R ...`) and it is ok.

Reply:
and I don't expect a user to pass a zpool that is already mounted. For me user can create its own complex setup but we will mount the zpool where we want.

Reply:
I expect the user to just `zpool create whatever vdev config`, which results in a mounted zpool

Reply:
I expected to have something not mounted. But if it is already mounted then the `pool_mounpoint` will not failed. It will give us the mount point of the pool so we will not import it.